File: replace_all_tool.py
import re
from pathlib import Path
from check_file_permission import check_file_permission

def replace_all(dir_path, pattern, replacement, config, regex=False):
    if not check_file_permission(dir_path, "write", config):
        raise PermissionError(f"Write access denied for {dir_path}")
    dir_path = Path(dir_path).resolve()
    results = []
    for file_path in dir_path.rglob("*"):
        if file_path.is_file() and check_file_permission(str(file_path), "write", config):
            with open(file_path, "r", encoding="utf-8") as f:
                content = f.read()
            new_content = re.sub(pattern, replacement, content) if regex else content.replace(pattern, replacement)
            with open(file_path, "w", encoding="utf-8") as f:
                f.write(new_content)
            results.append(f"Updated {file_path}")
    return "\n".join(results)

File: ts-tools/replace_all_tool.ts
/**
 * @license
 * Copyright 2025 Google LLC
 * SPDX-License-Identifier: Apache-2.0
 */

import * as fs from 'fs';
import * as path from 'path';
import { checkFilePermission } from './check_file_permission';

async function* walk(dir: string): AsyncGenerator<string> {
  for await (const d of await fs.promises.opendir(dir)) {
    const entry = path.join(dir, d.name);
    if (d.isDirectory()) yield* walk(entry);
    else if (d.isFile()) yield entry;
  }
}

export async function replaceAll(
  dirPath: string,
  pattern: string,
  replacement: string,
  config: any,
  regex: boolean = false,
): Promise<string> {
  await checkFilePermission(dirPath, 'write', config);
  const resolvedPath = path.resolve(dirPath);
  const results: string[] = [];
  for await (const p of walk(resolvedPath)) {
    await checkFilePermission(p, 'write', config);
    const content = await fs.promises.readFile(p, 'utf-8');
    const newContent = regex
      ? content.replace(new RegExp(pattern, 'g'), replacement)
      : content.replace(pattern, replacement);
    await fs.promises.writeFile(p, newContent, 'utf-8');
    results.push(`Updated ${p}`);
  }
  return results.join('\n');
}
